src/widgets/barchart/mod.rs

```diff
@@ -122,12 +120,39 @@ impl<'a> BarChart<'a> {
         self
     }
 
+    pub fn group_gap(mut self, gap: u16) -> BarChart<'a> {
+        self.group_gap = gap;
+        self
+    }
+
     pub fn style(mut self, style: Style) -> BarChart<'a> {
         self.style = style;
         self
     }
 }
 
+impl<'a> BarChart<'a> {
+    fn remove_unvisible_groups_and_bars(&mut self, mut width: u16) {
+        for group_index in 0..self.data.len() {
+            let n_bars = self.data[group_index].bars.len() as u16;
+            let bars_width = n_bars * self.bar_width + n_bars.saturating_sub(1) * self.bar_gap;
+
+            if width > bars_width {
+                width = width.saturating_sub(bars_width + self.group_gap);
+            } else {
+                let max_bars = (width + self.bar_gap) / (self.bar_width + self.bar_gap);
+                if max_bars == 0 {
+                    self.data.truncate(group_index);
+                } else {
+                    self.data[group_index].bars.truncate(max_bars as usize);
+                    self.data.truncate(group_index + 1);
+                }
+                break;
+            }
+        }
+    }
+}
+
 impl<'a> Widget for BarChart<'a> {
     fn render(mut self, area: Rect, buf: &mut Buffer) {
```

Comment:
This method started out large and got almost twice as big, which will make it twice as difficult to change in the future (likely more than twice as the number of paths through the method increases non linearly). If you were to describe the responsibilities of this method or describe the algorithm, would there be easy parts of it to refactor towards something easier to maintain?

It might be helpful to add some unit tests of the the new functionality so that any refactoring doesn't introduce bugs.

I haven't reviewed the code changes here yet, because I'd expect this to change pretty wildly if you refactor.

Reply:
In the latest commit I moved the code, which handles printing the bars to an extra function `print_bars`.

maybe it improves the readability a little bit. if it is a good idea, then I can squash commit it with the rest of the changes